### PR TITLE
Add appeals

### LINKED ISSUE #9791: Appeals Function
Just saw the new update for mastodon's moderation tool for putting accounts in limited mode or banning mode. I'd like to see a form available to folks who are in these modes to be able to submit an appeal request, to which a moderator/admin can review and consider either to deny, to remove their limited or banned mode, or to give them a certain time of being banned/limited as a smaller form of punishment.

### PR DESCRIPTION
**Pull request by Gargron**
Fix #9791

Allow users to view "strikes" (actions taken against an account) and to submit appeals through the form. Moderators are notified about the appeals and can choose to approve them to undo the strike, e.g. unsuspend a suspended account, or restore deleted posts. User is notified if their appeal is approved.

image
image

___

Funded through the BMBF (BMBF Förderkennzeichen: 01IS21S29)

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
I don't understand why you don't want to put the username right besides the display name like everywhere else here
image

**Comment by ClearlyClaire** [comment]
You can only approve appeals? … you can't reject them?

**Comment by Gargron** [comment]
To reject them you just do nothing

**Comment by ClearlyClaire** [comment]
Does this mean that if no moderator has report email enabled (which is probably a bad thing), moderators can't see whether there are appeals?

**Comment by Gargron** [comment]
Yes, it does mean that

**Comment by ClearlyClaire** [comment]
ok, that's probably a shortcoming then…

**Comment by Gargron** [comment]
Seems like it could be added later

**Comment by ClearlyClaire** [comment]
I'd rather be sure that a new moderation feature is complete and doesn't hinge on mail delivery being enabled and reliable for every involved party.

**Comment by Gargron** [comment]
Browsing appeals:

image

Visible on dashboard:

image

And on account pages:

image

**Comment by Gargron** [comment]
Strikes and appeals accessible on account settings page:

image

File: app/views/disputes/strikes/show.html.haml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        - if @strike.appealed?
+          %del= t(@strike.action, scope: 'user_mailer.warning.title')
+        - else
+          = t(@strike.action, scope: 'user_mailer.warning.title')
+    - if @strike.report && can?(:show, @strike.report)
+      .report-header__details__item
+        .report-header__details__item__header
+          %strong= t('disputes.strikes.associated_report')
+        .report-header__details__item__content
+          = link_to t('admin.reports.report', id: @strike.report.id), admin_report_path(@strike.report)
```
Comment (ClearlyClaire) [inline app/views/disputes/strikes/show.html.haml:92]: The styling for this one seems wrong:
image

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
First, thank you for addressing my previous concerns, namely:
- the display name being used without the username alongside (now, only the username is displayed), which could have been used for phishing-type attacks
- appeals not showing up anywhere besides in mail notifications (they can now be accessed from any place I would expect, and can be listed by moderators)

In addition to the inline comments, I think the feature can be useful and at least worth considering, **however** I'm worried about the signal-to-noise ratio and the amount of junk and abusive appeals admins will get from it. In that sense, I would prefer if:
- admins could opt out of the feature altogether (without making users think they can appeal) if it turns out to be too much trouble
- moderators could chose to not be notified of appeals while still notified of reports (so that work can be split differently etc.)

In addition, I have other concerns:
- **reverting post deletion will just cause issues** as by that time, post deletions are already federated (and it's a good thing for that to be near-immediate, as remote reporters would not understand the posts not being deleted), which is, to most implementations, irreversible (in order to prevent replay attacks etc.)
- I don't know if it's a good idea to notify users of rejected appeals
- I don't think it's good to list old rejected strikes in an otherwise-unrelated page in the account settings (for actionable strikes, it's a good thing); maybe strikes should even expire after a long while (even from moderator view) but I'm really not sure about that
- overall, I don't really know about the timing of the appeal process: I think replying immediately can be useful for obvious moderation errors, but detrimental when the dispute is founded; it also may make sense for someone to apologize and show improvements months later; but at the same time, some things, like suspensions, are irreversible by that time

File: app/controllers/admin/dashboard_controller.rb [hunk 2]
```diff
@@ -8,6 +8,7 @@ def index
       @pending_users_count   = User.pending.count
       @pending_reports_count = Report.unresolved.count
       @pending_tags_count    = Tag.pending_review.count
+      @pending_appeals_count = Appeal.pending.count
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/admin/dashboard_controller.rb:11]: If the table grows large enough. This is a page that is only visited by a few people so saving an index seems fine here.

File: app/controllers/admin/disputes/appeals_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    log_action :approve, @appeal
+    ApproveAppealService.new.call(@appeal, current_account)
+    redirect_to disputes_strike_path(@appeal.strike)
+  end
+
+  def reject
+    authorize @appeal, :approve?
+    log_action :reject, @appeal
+    @appeal.reject!(current_account)
+    UserMailer.appeal_rejected(@appeal.account.user, @appeal)
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/admin/disputes/appeals_controller.rb:23]: I think that if we have an explicit reject function, we might as well be upfront and notify people. Imagine you are appealing a copyright strike on YouTube and they wouldn't notify you about the final decision on that.

File: app/controllers/admin/disputes/appeals_controller.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  private
+
+  def filtered_appeals
+    Admin::AppealFilter.new(filter_params.with_defaults(status: 'pending')).results.includes(strike: :account)
+  end
+
+  def filter_params
+    params.slice(:page, *Admin::AppealFilter::KEYS).permit(:page, *Admin::AppealFilter::KEYS)
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/admin/disputes/appeals_controller.rb:34]: Just for consistency with other `filter_param` methods in other controllers.

File: app/controllers/auth/registrations_controller.rb [hunk 5]
```diff
@@ -123,6 +126,10 @@ def set_sessions
     @sessions = current_user.session_activations
   end
 
+  def set_strikes
+    @strikes = current_account.strikes.where(appealed_at: nil).latest
+  end
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/auth/registrations_controller.rb:131]: What do you call it when a strike has been successfully appealed? Is it "overruled"?

File: app/services/approve_appeal_service.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    @strike.touch(:appealed_at)
+  end
+
+  def undo_disable!
+    target_account.user.enable!
+  end
+
+  def undo_delete_statuses!
+    @strike.statuses.each(&:undiscard)
+  end
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/services/approve_appeal_service.rb:50]: I misread something in the code, you are correct.

File: app/controllers/admin/accounts_controller.rb [hunk 7]
```diff
@@ -146,7 +146,7 @@ def filtered_accounts
     end
 
     def filter_params
-      params.slice(*AccountFilter::KEYS).permit(*AccountFilter::KEYS)
+      params.slice(:page, *AccountFilter::KEYS).permit(:page, *AccountFilter::KEYS)
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/admin/accounts_controller.rb:149]: To be honest it's a mess currently, and I'm not sure having them in `filter_params` helps, you seem to need more code to filter it out, and you never access it through `filter_params`. I think I'd rather change the other controllers in a different way. But in another PR!

File: app/controllers/admin/dashboard_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/admin/dashboard_controller.rb:11]: This page is only visited by a few people but you want it to be fast.

File: app/helpers/admin/trends/statuses_helper.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        status.text.split("\n").first
+      else
+        Nokogiri::HTML(status.text).css('html > body > *').first&.text
+      end
+    end
+
+    return '' if text.blank?
+
+    html = Formatter.instance.send(:encode, text)
+    html = Formatter.instance.send(:encode_custom_emojis, html, status.emojis, prefers_autoplay?)
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/helpers/admin/trends/statuses_helper.rb:16]: Uncomfortable making that step now. Ideally Formatter needs a rewrite where it no longer relies on a status object being passed to it.

File: app/controllers/admin/accounts_controller.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/admin/accounts_controller.rb:149]: Actually, realized this is needed if we want batch actions to redirect us back to the same page. So it was a bug before.

File: app/controllers/admin/dashboard_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/admin/dashboard_controller.rb:11]: I insist on delaying it until at least a separate PR due to difficulties maintaining a clean schema.rb between PRs.

File: app/controllers/auth/registrations_controller.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/auth/registrations_controller.rb]: How is this solved?

File: app/views/auth/registrations/_status.html.haml [hunk 9]
```diff
@@ -1,22 +1,17 @@
...
-      %span.warning-hint= t('auth.status.confirming')
-      = link_to t('auth.didnt_get_confirmation'), new_user_confirmation_path
-    - elsif !@user.approved?
-      %span.warning-hint= t('auth.status.pending')
-    - elsif @user.account.moved_to_account_id.present?
-      %span.positive-hint= t('auth.status.redirecting_to', acct: @user.account.moved_to_account.acct)
-      = link_to t('migrations.cancel'), settings_migration_path
-    - else
-      %span.positive-hint= t('auth.status.functional')
+= render partial: 'account_warning', collection: @strikes
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/views/auth/registrations/_status.html.haml:15]: Better now, but I wouldn't want to encourage someone to appeal a strike to make it disappear. It also doesn't make sense when the user cannot appeal to the strike anymore. The check should probably be based on `created_at`.

[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/views/auth/registrations/_status.html.haml:15]: I guess it still does, but what about strikes for deleted content?

[the remark this answers is not in the document]
Comment (Gargron) [reply app/views/auth/registrations/_status.html.haml:15]: I'm still of the opinion that past warnings should be accessible. And even though you may be unsilenced separately, this is still history that happened.

**Comment by nightpool** [comment]
I think being able to reverse post deletions is an important moderation tool, and I would prefer that we do the work to support that (although for scope reasons it probably should be separate from this PR? so I'm okay with having undeletions be slightly weird or broken until then)

**Comment by nightpool** [comment]
that seems fine to me as an interim step as long as we have a plan to
address it in the future.

**Comment by Gargron** [comment]
So I should undo my undo of the undo feature?

**Comment by ClearlyClaire** [comment]
I'd rather the interim step being the undo not being available until we can federate it.

**Comment by Gargron** [comment]
I think I addressed everything here for now.

**Comment by ClearlyClaire** [comment]
I'm still unconvinced by the “show any strike forever unless it has been explicitly overruled” thing. Like I said, you probably don't want to be reminded forever that your moderators deleted 3 times 2 toots of you months ago.

Otherwise, maybe make it explicit that deleted toots are not restored when the decision is overruled, until we have a solution for this.

Finally, I still think admins should be able to disable the feature entirely, but mods being able to opt out of appeal emails is already a good step, and I guess we can first see how it turns out on real instances.

**Comment by Gargron** [comment]
This can be adjusted later, even before next release. I’m sure once i run it live I’ll find things to improve. I don’t think this should prevent a merge.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by Gargron** [comment]
Thank you!

**Pull request merged**